File: bindings/lua/src/uhppote.lua
local uhppote = {}
local encode = require("src/encode")
local decode = require("src/decode")
local ut0311 = require("src/ut0311")
local structs = require("src/structs")

local function resolve(controller)
    if type(controller) == 'number' then
        return structs.controller(controller, "", "udp")
    end

    if type(controller) == 'table' and getmetatable(controller) ~= nil and getmetatable(controller).__type == 'structs.controller' then
        return controller
    end

    return structs.controller(0, "", "udp")
end

function uhppote.set_bind_address(address)
    ut0311.set_bind_address(address)
end

function uhppote.set_broadcast_address(address)
    ut0311.set_broadcast_address(address)
end

function uhppote.set_listen_address(address)
    ut0311.set_listen_address(address)
end

function uhppote.set_debug(debug)
    ut0311.set_debug(debug)
end

function uhppote.get_all_controllers()
    local request = encode.get_controller_request(0)
    local replies = ut0311.broadcast(request)

    if not replies then
        error("no response")
    end

    local list = {}
    for k, reply in ipairs(replies) do
        local response = decode.get_controller_response(reply)

        table.insert(list, response)
    end

    return list
end

function uhppote.listen(handler, onerror)
    local f = function(packet)
        local ok, event = xpcall(function()
            return decode.event(packet)
        end, onerror)
        if ok then
            handler(event)
        end
    end

    ut0311.listen(f)
end

{{range .model.functions}}
{{- template "function" . -}}
{{end}}

{{define "function"}}
function uhppote.{{snakeCase .name}}({{template "args" .args}})
    local c = resolve(controller)
    local request = encode.{{snakeCase .request.name}}(c['controller']{{template "params" slice .args 1}})
    local reply = ut0311.send(c, request)
    {{if .response}}
    if not reply then
        error("no response")
    end

    return decode.{{snakeCase .response.name}}(reply)
    {{- else}}
    return nil
    {{- end}}
end
{{end}}

return uhppote

File: bindings/python/uhppote.py
from collections import namedtuple

import encode
import decode
import net

Controller = namedtuple('Controller', 'controller address transport')

class Uhppote:
    def __init__(self, bind='0.0.0.0', broadcast='255.255.255.255:60000', listen="0.0.0.0:60001", debug=False):
        self._net = net.Net(bind, broadcast, listen, debug)

    def get_all_controllers(self):
        request = encode.get_controller_request(0)
        replies = self._net.broadcast(request)

        list = []
        for reply in replies:
            list.append(decode.get_controller_response(reply))

        return list

    def listen(self, onEvent):
        self._net.listen(lambda packet: self.events(packet, onEvent))
        return None


    def events(self, packet, onEvent):
        onEvent(decode.event(packet))
    
{{range .model.functions}}
{{- template "function" . -}}
{{end}}

{{define "function"}}
    def {{snakeCase .name}}(self, {{template "args" .args}}):
        {{if .response}}request = encode.{{snakeCase .request.name}}(controller.controller{{template "params" slice .args 1}})
        reply = self._net.send(controller, request)

        if reply != None:
            return decode.{{snakeCase .response.name}}(reply)
            
        return None
        {{- else}}
        request = encode.{{snakeCase .request.name}}(controller.controller{{template "params" slice .args 1}})
        self._net.send(controller, request)

        return True
        {{end}}
{{end}}
